### ISSUE TITLE
Build failures on Ubuntu 14.04 with clang 3.4

### ISSUE THREAD

**Issue by ms705** [Microsoft/cpprestsdk#61]
On a fully up-to-date Ubuntu 14.04 LTS installation, I get the following error when building cpprestsdk (specifically, `test_runner`) with clang instead of GCC:

```shell
$ cmake .. -DCMAKE_BUILD_TYPE=Release -DCMAKE_CXX_COMPILER=clang++ -DCMAKE_C_COMPILER=clang
[...]
-- The C compiler identification is Clang 3.4.0
-- The CXX compiler identification is Clang 3.4.0
-- Check for working C compiler: /usr/bin/clang
-- Check for working C compiler: /usr/bin/clang -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/clang++
-- Check for working CXX compiler: /usr/bin/clang++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Boost version: 1.55.0
-- Found the following Boost libraries:
--   random
--   chrono
--   system
--   thread
--   regex
--   filesystem
-- Looking for include file pthread.h
-- Looking for include file pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE
-- Found OpenSSL: /usr/lib/x86_64-linux-gnu/libssl.so;/usr/lib/x86_64-linux-gnu/libcrypto.so (found suitable version "1.0.1f", minimum required is "1.0.0")
-- Setting clang options
-- websocketpp not found, using the embedded version
-- Added test library httpclient_test
-- Added test library httplistener_test
-- Added test library json_test
-- Added test library pplx_test
-- Added test library streams_test
-- Added test library uri_test
-- Added test library utils_test
-- Added test library websocketclient_test
-- Configuring done
-- Generating done
-- Build files have been written to: /src/cpprestsdk-git/Release/build.release
$
$ make
[  0%] Building CXX object src/CMakeFiles/cpprest.dir/http/client/http_client_msg.cpp.o
[...]
Linking CXX shared library ../Binaries/libcpprest.so
[ 19%] Built target cpprest
[...]
[ 32%] Building CXX object tests/common/UnitTestpp/CMakeFiles/unittestpp.dir/src/Posix/TimeHelpers.cpp.o
Linking CXX shared library ../../../Binaries/libunittestpp.so
[ 32%] Built target unittestpp
[ 33%] Building CXX object tests/common/utilities/CMakeFiles/common_utilities.dir/os_utilities.cpp.o
[ 34%] Building CXX object tests/common/utilities/CMakeFiles/common_utilities.dir/stdafx.cpp.o
Linking CXX shared library ../../../Binaries/libcommon_utilities.so
[ 34%] Built target common_utilities
[...]
[ 35%] Building CXX object tests/common/TestRunner/CMakeFiles/test_runner.dir/test_module_loader.cpp.o
[...]
Linking CXX executable ../../../Binaries/test_runner
CMakeFiles/test_runner.dir/test_runner.cpp.o: In function `boost::filesystem::path::path<boost::filesystem::directory_entry>(boost::filesystem::directory_entry const&, boost::enable_if<boost::filesystem::path_traits::is_pathable<boost::decay<boost::filesystem::directory_entry>::type>, void>::type*)':
/src/cpprestsdk-git/Release/tests/common/TestRunner/test_runner.cpp:(.text._ZN5boost10filesystem4pathC2INS0_15directory_entryEEERKT_PNS_9enable_ifINS0_11path_traits11is_pathableINS_5decayIS4_E4typeEEEvE4typeE[_ZN5boost10filesystem4pathC2INS0_15directory_entryEEERKT_PNS_9enable_ifINS0_11path_traits11is_pathableINS_5decayIS4_E4typeEEEvE4typeE]+0xc3): undefined reference to `boost::filesystem::path_traits::dispatch(boost::filesystem::directory_entry const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&, std::__1::codecvt<wchar_t, char, __mbstate_t> const&)'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [Binaries/test_runner] Error 1
make[1]: *** [tests/common/TestRunner/CMakeFiles/test_runner.dir/all] Error 2
make: *** [all] Error 2
```

I understand that the [documentation](https://github.com/Microsoft/cpprestsdk/wiki/How-to-build-for-Linux) recommends using `g++4.8` on Linux, but I had hoped that clang would work anyway.

My clang version is `Ubuntu clang version 3.4-1ubuntu3`, and my Boost version is 1.55, with the `libboost-filesystem1.55-dev` package installed. Using `g++4.8` as CXX, the build succeeds without any problems.

**Comment by ras0219-msft** [comment]
Try using the development branch. I believe this is a bug where we forgot to include `boost_filesystem` on the link line for the tests. The relevant commit is: https://github.com/Microsoft/cpprestsdk/commit/95478a4c83a6b02456deede288c2496f813e7d9d

That said, I would have expected the same behavior from gcc and clang here...

**Comment by AdeMiller** [comment]
I'm not even getting that far.
```
ade@linux-ycnz:~/src/casablanca/Release/build.release> CXX=clang cmake .. -DCMAKE_BUILD_TYPE=Release -DBOOST_ROOT=../../boost
-- Boost version: 1.60.0
-- Found the following Boost libraries:
--   random
--   chrono
--   system
--   thread
--   regex
--   filesystem
-- Setting clang options
-- websocketpp not found, using the embedded version
-- Added test library httpclient_test
-- Added test library httplistener_test
-- Added test library json_test
-- Added test library pplx_test
-- Added test library streams_test
-- Added test library uri_test
-- Added test library utils_test
-- Added test library websocketclient_test
-- Configuring done
-- Generating done
-- Build files have been written to: /home/ade/src/casablanca/Release/build.release
ade@linux-ycnz:~/src/casablanca/Release/build.release> 
ade@linux-ycnz:~/src/casablanca/Release/build.release> make
Scanning dependencies of target cpprest
[  1%] Building CXX object src/CMakeFiles/cpprest.dir/http/client/http_client_msg.cpp.o
[  1%] Building CXX object src/CMakeFiles/cpprest.dir/http/client/x509_cert_utilities.cpp.o
[  2%] Building CXX object src/CMakeFiles/cpprest.dir/http/common/http_helpers.cpp.o
/home/ade/src/casablanca/Release/src/http/common/http_helpers.cpp:275:16: error: redundant move in return statement
      [-Werror,-Wredundant-move]
        return std::move(src);
               ^
/home/ade/src/casablanca/Release/src/http/common/http_helpers.cpp:275:16: note: remove std::move call here
        return std::move(src);
               ^~~~~~~~~~   ~
/home/ade/src/casablanca/Release/src/http/common/http_helpers.cpp:316:16: error: redundant move in return statement
      [-Werror,-Wredundant-move]
        return std::move(src);
               ^
/home/ade/src/casablanca/Release/src/http/common/http_helpers.cpp:316:16: note: remove std::move call here
        return std::move(src);
               ^~~~~~~~~~   ~
/home/ade/src/casablanca/Release/src/http/common/http_helpers.cpp:327:12: error: redundant move in return statement
      [-Werror,-Wredundant-move]
    return std::move(src);
           ^
/home/ade/src/casablanca/Release/src/http/common/http_helpers.cpp:327:12: note: remove std::move call here
    return std::move(src);
           ^~~~~~~~~~   ~
3 errors generated.
src/CMakeFiles/cpprest.dir/build.make:110: recipe for target 'src/CMakeFiles/cpprest.dir/http/common/http_helpers.cpp.o' failed
make[2]: *** [src/CMakeFiles/cpprest.dir/http/common/http_helpers.cpp.o] Error 1
CMakeFiles/Makefile2:85: recipe for target 'src/CMakeFiles/cpprest.dir/all' failed
make[1]: *** [src/CMakeFiles/cpprest.dir/all] Error 2
Makefile:138: recipe for target 'all' failed
make: *** [all] Error 2
```
I'll investigate further tomorrow. I'm just wondering if clang should work and this is bug or clang really isn't supported.

Thx.

**Comment by expandrive** [comment]
This also occurs on OS X using clang, the 2.8 release build fails

**Comment by ras0219-msft** [comment]
Could you elaborate on what specific versions of boost and xcode and clang you're using? What specific issues are you seeing when compiling the development branch?

**Comment by ms705** [comment]
Hi @ras0219-msft,

Thanks for the response! The immediate issue I experienced above is indeed fixed by 95478a4c83a6b02456deede288c2496f813e7d9d. However, a similar issue remains both with current `master` and `develop` branches, and, in fact, both clang and g++.

The linker error is now:
```
Linking CXX executable ../../../Binaries/test_runner
cd /tmp/cpprestsdk/Release/tests/common/TestRunner && /usr/bin/cmake -E cmake_link_script CMakeFiles/test_runner.dir/link.txt --verbose=1
/usr/bin/clang++-3.5    -stdlib=libc++ -Wno-return-type-c-linkage -Wno-unneeded-internal-declaration -std=c++11 -fno-strict-aliasing -g    CMakeFiles/test_runner.dir/test_runner.cpp.o CMakeFiles/test_runner.dir/test_module_loader.cpp.o  -o ../../../Binaries/test_runner -rdynamic -lboost_random -lboost_chrono -lboost_system -lboost_thread -lboost_regex -lboost_filesystem -lpthread ../../../Binaries/libunittestpp.so -ldl -Wl,-rpath,/tmp/cpprestsdk/Release/Binaries 
CMakeFiles/test_runner.dir/test_runner.cpp.o: In function `path<boost::filesystem::directory_entry>':
/usr/include/boost/filesystem/path.hpp:139: undefined reference to `boost::filesystem::path_traits::dispatch(boost::filesystem::directory_entry const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&, std::__1::codecvt<wchar_t, char, __mbstate_t> const&)'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

Boost version is 1.54, g++ 4.8.4, clang 3.5.0-4ubuntu2~trusty2, on Ubuntu 14.04.4.

I'll try to dig in and see what's wrong here -- #42 seems related, but was allegedly fixed by 95478a4c83a6b02456deede288c2496f813e7d9d. Strange!
